src/redisgears_python.c

```diff
@@ -65,145 +65,367 @@ static void* RedisGearsPy_PyCallbackDeserialize(FlatExecutionPlan* fep, Gears_Bu
+
 typedef struct PythonSessionCtx{
     size_t refCount;
     char sessionId[ID_LEN];
     char sessionIdStr[STR_ID_LEN];
     PyObject* globalsDict;
-    char** requirementList;
+    PythonRequirementCtx** requirements;
 }PythonSessionCtx;
 
-static void* PythonSessionCtx_ShellowCopy(void* arg){
-    PythonSessionCtx* session = arg;
-    ++session->refCount;
-    return session;
-}
+static bool PythonRequirementCtx_DownloadRequirement(PythonRequirementCtx* req){
+#define RETRY 3
+#define RETRY_SLEEP_IN_SEC 1
+    int exitCode;
+    for(size_t i = 0 ; i < RETRY; ++i){
+        exitCode = ExecCommand(NULL, "/bin/bash -c \"source %s/bin/activate;cd %s;python -m pip wheel %s\"", venvDir, req->basePath, req->name);
+        if(exitCode != 0){
+            sleep(RETRY_SLEEP_IN_SEC);
+            continue;
+        }
+        break;
+    }
+    if(exitCode != 0){
+        return false;
+    }
 
-static bool PythonSessionCtx_IsRequirementInstalled(const char* requirement){
-    for(size_t i = 0 ; i < array_len(requitmentsCache) ; ++i){
-        if(strcmp(requirement, requitmentsCache[i]) == 0){
-            return true;
+    // fills the wheels array
+    DIR *dr = opendir(req->basePath);
+    assert(dr);
+    struct dirent *de;
+    while ((de = readdir(dr))){
+        if (strcmp(de->d_name, ".") == 0 || strcmp(de->d_name, "..") == 0){
+            continue;
         }
+        char* c = de->d_name;
+        req->wheels = array_append(req->wheels, RG_STRDUP(de->d_name));
     }
-    return false;
+    closedir(dr);
+
+    return true;
 }
 
-static void PythonSessionCtx_FreeRequirementsList(char** requirementsList){
-    for(size_t i = 0 ; i < array_len(requirementsList) ; ++i){
-        RG_FREE(requirementsList[i]);
+static bool PythonRequirementCtx_InstallRequirement(PythonRequirementCtx* req){
+    char* filesInDir = array_new(char, 10);
+    for(size_t i = 0 ; i < array_len(req->wheels) ; ++i){
+        char* c = req->wheels[i];
+        while(*c){
+            filesInDir = array_append(filesInDir, *c);
+            ++c;
+        }
+        filesInDir = array_append(filesInDir, ' ');
     }
-    array_free(requirementsList);
+    filesInDir[array_len(filesInDir) - 1] = '\0';
+
+    int exitCode = ExecCommand(NULL, "/bin/bash -c \"source %s/bin/activate;cd %s;python -m pip install %s\"", venvDir, req->basePath, filesInDir);
+    array_free(filesInDir);
+    if(exitCode != 0){
+        return false;
+    }
+    return true;
 }
 
-static PythonSessionCtx* PythonSessionCtx_GetOrCreate(char* id, char** requirementsList){
-    if(id){
-        PythonSessionCtx* session = Gears_dictFetchValue(SessionsDict, id);
-        if(session){
-            PythonSessionCtx_FreeRequirementsList(requirementsList);
-            return PythonSessionCtx_ShellowCopy(session);
+static void PythonRequirementCtx_Free(PythonRequirementCtx* reqCtx){
+    if(--reqCtx->refCount){
+        return;
+    }
+
+    ExecCommand(NULL, "rm -rf %s", reqCtx->basePath);
```

Comment:
reqCtx->basePath == "/" scares me.

Reply:
What do you suggest?

Reply:
From looking at the construction of basePath, it looks like we should check that it doesn't match /+. Maybe do it in a function and call it in creation site and rm -rf site just to be sure we don't rack the system.